File: harness/harness/security.py
"""Bash command allowlist — PreToolUse hook for sandbox defense-in-depth."""

import re
import shlex
from claude_agent_sdk import (
    HookJSONOutput,
    PreToolUseHookInput,
    HookContext,
)

# Commands the generator/evaluator agents are allowed to run
ALLOWED_COMMANDS = {
    "ls", "cat", "head", "tail", "wc", "grep", "find",
    "cp", "mv", "mkdir", "chmod", "pwd", "echo", "printf",
    "npm", "npx", "node", "tsc", "eslint",
    "git", "touch", "rm", "sort", "uniq", "diff",
    "curl", "jq", "sed", "awk",
    "python3", "pip",
}

# Commands that need extra validation
RESTRICTED_COMMANDS = {
    "rm": lambda args: not any(a in args for a in ["-rf /", "-rf ~", "--no-preserve-root"]),
    "chmod": lambda args: "+x" in args or "755" in args or "644" in args,
    "curl": lambda _: True,  # Allow but could restrict domains later
}


def _extract_base_commands(command: str) -> list[str]:
    """Extract base command names from a compound shell command."""
    # Split on shell operators
    parts = re.split(r'[|;&]|&&|\|\|', command)
    commands = []
    for part in parts:
        part = part.strip()
        if not part:
            continue
        try:
            tokens = shlex.split(part)
        except ValueError:
            # Unparseable — fail closed
            return ["__UNPARSEABLE__"]
        if tokens:
            # Skip env var assignments at the start
            cmd = tokens[0]
            for t in tokens:
                if "=" in t and t.index("=") > 0:
                    continue
                cmd = t
                break
            commands.append(cmd)
    return commands


async def bash_security_hook(
    input_data: PreToolUseHookInput,
    tool_use_id: str,
    context: HookContext,
) -> HookJSONOutput:
    """Validate bash commands against the allowlist."""
    tool_input = input_data.get("input", {})
    command = tool_input.get("command", "")

    if not command:
        return {"decision": "deny", "reason": "Empty command"}

    base_commands = _extract_base_commands(command)

    for cmd in base_commands:
        # Get just the binary name (strip path)
        cmd_name = cmd.split("/")[-1]

        if cmd_name == "__UNPARSEABLE__":
            return {
                "decision": "deny",
                "reason": f"Could not parse command: {command}",
            }

        if cmd_name not in ALLOWED_COMMANDS:
            return {
                "decision": "deny",
                "reason": f"Command '{cmd_name}' is not in the allowlist",
            }

        # Extra validation for restricted commands
        if cmd_name in RESTRICTED_COMMANDS:
            if not RESTRICTED_COMMANDS[cmd_name](command):
                return {
                    "decision": "deny",
                    "reason": f"Command '{cmd_name}' failed safety check: {command}",
                }

    return {"decision": "allow"}

File: src/engine/security.ts
/**
 * Bash command allowlist — PreToolUse hook for sandbox defense-in-depth.
 * Ported from salazar/salazar/security.py
 */

const ALLOWED_COMMANDS = new Set([
  "ls", "cat", "head", "tail", "wc", "grep", "find",
  "cp", "mv", "mkdir", "chmod", "pwd", "echo", "printf",
  "npm", "npx", "node", "tsc", "eslint",
  "git", "touch", "rm", "sort", "uniq", "diff",
  "curl", "jq", "sed", "awk",
]);

/**
 * Split a compound shell command on pipe/semicolon/ampersand operators
 * and return the base command name for each part.
 */
function extractBaseCommands(command: string): string[] {
  // Split on |, ;, &, &&, ||
  const parts = command.split(/\|{1,2}|;|&{1,2}/);
  const commands: string[] = [];

  for (const part of parts) {
    const trimmed = part.trim();
    if (!trimmed) continue;

    // Tokenize on whitespace (simple split — sufficient for allowlist check)
    const tokens = trimmed.split(/\s+/);
    if (tokens.length === 0) continue;

    // Skip leading env var assignments (FOO=bar)
    let cmd = tokens[0];
    for (const token of tokens) {
      // An env var assignment has '=' somewhere after position 0
      const eqIdx = token.indexOf("=");
      if (eqIdx > 0) {
        // This token is an assignment — skip it, keep looking
        continue;
      }
      cmd = token;
      break;
    }

    commands.push(cmd);
  }

  return commands;
}

/**
 * PreToolUse hook that validates bash commands against the allowlist.
 * Compatible with Agent SDK HookCallback signature.
 */
export async function bashSecurityHook(input: any): Promise<Record<string, unknown>> {
  // Agent SDK delivers the command at input.input.command or input.tool_input.command
  const toolInput = input?.input ?? input?.tool_input ?? {};
  const command: string = toolInput?.command ?? "";

  if (!command.trim()) {
    return { decision: "deny", reason: "Empty command" };
  }

  const baseCommands = extractBaseCommands(command);

  if (baseCommands.length === 0) {
    return { decision: "deny", reason: "Could not extract any commands" };
  }

  for (const cmd of baseCommands) {
    // Strip path prefix — get just the binary name
    const cmdName = cmd.split("/").at(-1) ?? cmd;

    if (!ALLOWED_COMMANDS.has(cmdName)) {
      return {
        decision: "deny",
        reason: `Command '${cmdName}' is not in the allowlist`,
      };
    }
  }

  return { decision: "allow" };
}
